subprojects/jacoco/src/main/java/org/gradle/testing/jacoco/plugins/JacocoPluginExtension.java

```diff
@@ -122,38 +121,55 @@ public void setReportsDir(File reportsDir) {
         final String taskName = task.getName();
         LOGGER.debug("Applying Jacoco to " + taskName);
         final JacocoTaskExtension extension = task.getExtensions().create(TASK_EXTENSION_NAME, JacocoTaskExtension.class, objects, agent, task);
-        extension.setDestinationFile(providerFactory.provider(new Callable<File>() {
+        extension.setDestinationFile(providers.provider(new Callable<File>() {
             @Override
             public File call() {
-                return fileOperations.file(projectLayout.getBuildDirectory().get().getAsFile() + "/jacoco/" + taskName + ".exec");
+                return layout.getBuildDirectory().file("jacoco/" + taskName + ".exec").get().getAsFile();
             }
         }));
 
         task.getJvmArgumentProviders().add(new JacocoAgent(extension));
-        task.doFirst(new Action<Task>() {
-            @Override
-            public void execute(Task task) {
-                if (extension.isEnabled() && extension.getOutput() == JacocoTaskExtension.Output.FILE) {
-                    // Delete the coverage file before the task executes, so we don't append to a leftover file from the last execution.
-                    // This makes the task cacheable even if multiple JVMs write to same destination file, e.g. when executing tests in parallel.
-                    // The JaCoCo agent supports writing in parallel to the same file, see https://github.com/jacoco/jacoco/pull/52.
-                    File coverageFile = extension.getDestinationFile();
-                    if (coverageFile == null) {
-                        throw new GradleException("JaCoCo destination file must not be null if output type is FILE");
-                    }
-                    fileOperations.delete(coverageFile);
-                }
-            }
-        });
+        task.doFirst(new JacocoOutputCleanupTestTaskAction(
+            fs,
+            providers.provider(() -> extension.isEnabled() && extension.getOutput() == JacocoTaskExtension.Output.FILE),
+            providers.provider(extension::getDestinationFile)
+        ));
 
         // Do not cache the task if we are not writing execution data to a file
-        task.getOutputs().doNotCacheIf("JaCoCo configured to not produce its output as a file", new Spec<Task>() {
-            @Override
-            public boolean isSatisfiedBy(Task element) {
-                // Do not cache Test task if Jacoco doesn't produce its output as files
-                return extension.isEnabled() && extension.getOutput() != JacocoTaskExtension.Output.FILE;
+        Provider<Boolean> doNotCachePredicate = providers.provider(() ->
```

Comment:
💄 same predicate is used above in `task.doFirst`, could reuse.

Reply:
It's not the same predicate

Reply:
Got it.